### PR TITLE
Implement radix tree

### PR DESCRIPTION
**Pull request by platfowner**
Change summary:
- Implement radix tree (w/ radix node)
- Add isHexString(), toHexString() util functions

Todos:
- Implement proof hash methods
- Integrate with StateNode (state-node.js) 

Related issues: 
https://github.com/ainblockchain/ain-blockchain/issues/519
https://github.com/ainblockchain/ain-blockchain/issues/517

### CODE REVIEW

File: db/radix-tree.js [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+      // Case 3: Has a child with matching label suffix.
+      curNode = child;
+      labelIndex += 1 + child.getLabelSuffix().length;
+    }
+
+    return curNode;
+  }
+
+  getFromTree(label) {
```
Comment (liayoo) [inline db/radix-tree.js:101]: suggestion) how about a note that says it's getting the StateNode? or maybe include it in the function name, like getStateNodeFromTree?

File: db/radix-node.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
+const logger = require('../logger')('RADIX_NODE');
+
+class RadixNode {
+  constructor() {
+    this.stateNode = null;
+    this.labelRadix = '';
+    this.labelSuffix = '';
+    this.parent = null;
+    this.childMap = new Map();
```
Comment (liayoo) [inline db/radix-node.js:9]: Could you make it a little clearer that this is a childMap composed of RadixNode?

File: db/radix-tree.js [hunk 3]
```diff
@@ -0,0 +1,241 @@
+const logger = require('../logger')('RADIX_TREE');
+
+const CommonUtil = require('../common/common-util');
+const RadixNode = require('./radix-node');
+
+class RadixTree {
+  constructor() {
+    this.root = new RadixNode();
+    this.childMap = new Map();
```
Comment (liayoo) [inline db/radix-tree.js:9]: Could you make it a little clearer that this is a childMap composed of StateNode?

File: db/radix-tree.js [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  getFromTree(label) {
+    const hexLabel = RadixTree.toHexLabel(label);
+    const node = this.getNodeForReading(hexLabel);
+    if (node === null) {
+      return null;
+    }
+    return node.getStateNode();
+  }
+
+  getFromMap(label) {
```
Comment (liayoo) [inline db/radix-tree.js:110]: ditto

**Review by liayoo — COMMENTED** [review]
Thanks for the nice implementation! It seems a little ambiguous at times whether a node / child is a StateNode or a RadixNode, so I left a couple of notes on that.. Let me know if I misunderstood anything!

File: db/radix-node.js (hunk 2 above)
Comment (platfowner) [reply db/radix-node.js]: Renamed it: -> 'radixChildMap'. PTAL~

File: db/radix-tree.js (hunk 3 above)
Comment (platfowner) [reply db/radix-tree.js]: Renamed: -> 'stateNodeMap'. PTAL~

File: db/radix-tree.js (hunk 1 above)
Comment (platfowner) [reply db/radix-tree.js]: All external methods of RadixTree is about state nodes. Addressed it with:
- Added underscore (_) prefix to all internal methods
- Rename methods: '_getRadixNodeForReading', '_getRadixNodeForWriting'
- Add jsdoc comment to RadixTree class

PTAL~

File: db/radix-tree.js (hunk 4 above)
Comment (platfowner) [reply db/radix-tree.js]: ditto~

**Comment by platfowner** [comment]
All comments addressed, PTAL~ @liayoo

**Comment by liayoo** [comment]
Thanks for the updates! Looks good! 👍 I just have one more question about the design - how do the radix trees fit in with StateManager? Will it be changed to manage radix trees instead of state node trees?

**Comment by platfowner** [comment]
> Thanks for the updates! Looks good! 👍 I just have one more question about the design - how do the radix trees fit in with StateManager? Will it be changed to manage radix trees instead of state node trees?

My idea is to replace StateNode.childMap with RadixTree (this is why RadixTree has 1-to-1 mapped APIs with Map()) as described in AIP-025, so each state node will have one instance of RadixTree, which means they won't be managed in a center (and they have nothing to do with state versions, which is managed by StateManager - I have a plan to give a new name StateVersionManager).

**Comment by liayoo** [comment]
> > Thanks for the updates! Looks good! 👍 I just have one more question about the design - how do the radix trees fit in with StateManager? Will it be changed to manage radix trees instead of state node trees?
> 
> My idea is to replace StateNode.childMap with RadixTree (this is why RadixTree has 1-to-1 mapped APIs with Map()) as described in AIP-025, so each state node will have one instance of RadixTree, which means they won't be managed in a center (and they have nothing to do with state versions, which is managed by StateManager - I have a plan to give a new name StateVersionManager).

Ah, now I get it! That's brilliant. Thanks for the explanation!

File: db/radix-tree.js [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+    // 1. Set in the radix tree.
+    this._setInTree(label, stateNode);
+    // 2. Set in the hash map.
+    this._setInMap(label, stateNode);
+  }
+
+  _hasInTree(label) {
+    const hexLabel = RadixTree._toHexLabel(label);
+    const node = this._getRadixNodeForReading(hexLabel);
+    if (node === null) {
```
Comment (liayoo) [inline db/radix-tree.js:148]: minor) how about `return node !== null` ?

File: db/radix-tree.js [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    const hexLabel = RadixTree._toHexLabel(label);
+    const node = this._getRadixNodeForReading(hexLabel);
+    if (node === null || !node.hasStateNode()) {
+      logger.error(`[${LOG_HEADER}] Deleting a non-existing child with label: ${label}.`);
+      // Does nothing.
+      return false;
+    }
+    node.resetStateNode();
+    if (node.numChildren() === 1 && node.hasParent()) {
+      this._mergeToChild(node);
```
Comment (liayoo) [inline db/radix-tree.js:203]: How about return this._mergeToChild(node) ?

File: db/radix-tree.js [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    } else if (node.numChildren() === 0) {
+      if (!node.hasParent()) {
+        logger.error(`[${LOG_HEADER}] Deleting a child without parent with label: ${label}.`);
+        // Does nothing.
+        return false;
+      }
+      const parent = node.getParent();
+      parent.deleteChild(node.getLabelRadix());
+      if (parent.numChildren() === 1 && !parent.hasStateNode() && parent.hasParent()) {
+        this._mergeToChild(parent);
```
Comment (liayoo) [inline db/radix-tree.js:213]: ditto

File: unittest/rule-util.test.js [hunk 8]
```diff
@@ -171,14 +171,44 @@ describe("RuleUtil", () => {
...
+
   describe("isValidHash", () => {
     it("when invalid input", () => {
       expect(util.isValidHash(true)).to.equal(false);
       expect(util.isValidHash(false)).to.equal(false);
       expect(util.isValidHash(0)).to.equal(false);
       expect(util.isValidHash(10)).to.equal(false);
-      expect(util.isDict(null)).to.equal(false);
-      expect(util.isDict(undefined)).to.equal(false);
+      expect(util.isValidHash(null)).to.equal(false);
```
Comment (liayoo) [inline unittest/rule-util.test.js:210]: Oops! Thanks for catching this!

**Review by liayoo — APPROVED** [review]
LGTM with some minor comments. Thanks!

File: db/radix-tree.js (hunk 5 above)
Comment (platfowner) [reply db/radix-tree.js]: Done, thanks!

File: db/radix-tree.js (hunk 6 above)
Comment (platfowner) [reply db/radix-tree.js]: Done, thanks!

File: db/radix-tree.js (hunk 7 above)
Comment (platfowner) [reply db/radix-tree.js]: ditto~

**Pull request merged**